File: Autocomplete_Brute.py
# Time complexity : O(n log k) --> k = 3, n = strings in hashmap
# Space complexity : O(n) --> frequency map
# Brute force solution
class TrieNode:
    def __init__(self,):
        self.children = [None for _ in range(256)]
        self.strList = []
        
        
        
class AutocompleteSystem:
    # custom class for Heap
    class Distance:
        # passing the string and the Hashmap to the function
        def __init__(self, string, strs):
            self.string = string
            self.strs = strs
        # compare the two strings
        def __lt__(self, other):
            # if frequency is same, return the lexo string
            if self.strs[self.string] == self.strs[other.string]:
                return self.string > other.string
            # return the string with higher freq
            return self.strs[self.string] < self.strs[other.string]
        def __gt__(self, other):
            # if frequency is same, return the lexo string
            if self.strs[self.string] == self.strs[other.string]:
                return self.string < other.string
            # return the string with higher freq
            return self.strs[self.string] > self.strs[other.string]
    def insert(self, sentence):
            curr = self.root
            for i in range(len(sentence)):
                ch = sentence[i]
                if curr.children[ord(ch) - ord(' ')] == None:
                    curr.children[ord(ch) - ord(' ')] = TrieNode()
                curr = curr.children[ord(ch) - ord(' ')]
                curr.strList.append(sentence)
        
    def search(self, prefix):
        curr = self.root
        for i in range(len(prefix)):
            ch = prefix[i]
            if curr.children[ord(ch) - ord(' ')] == None:
                return []
            curr = curr.children[ord(ch) - ord(' ')]
        return curr.strList
        
    def __init__(self, sentences: List[str], times: List[int]):
        self.sentences = sentences
        self.times = times
        self.freq = {}
        self.input_str = ''
        self.root = TrieNode()
        
        # building Hashmap with the frequency
        for i in range(len(self.sentences)):
            if self.sentences[i] in self.freq:
                self.freq[self.sentences[i]] = self.freq[self.sentences[i]] + times[i]
            else:
                self.freq[self.sentences[i]] = times[i]
                self.insert(self.sentences[i])
            

    def input(self, c: str) -> List[str]:
        # if the input is # then we check for the string in Hashmap and increase it's frequency is present
        # else make it 1, and return an empty string
        if c == '#':
            search_str = self.input_str
            if search_str not in self.freq:
                self.insert(search_str)
                self.freq[search_str] = 1
            else:
                self.freq[search_str] += 1
            
            self.input_str = ''
            return []
        self.input_str += c
        minHeap = []
        prefix = self.input_str
        starts_with = self.search(prefix)
        for st in starts_with:
            heapq.heappush(minHeap, self.Distance(st,self.freq))
            if len(minHeap) > 3:
                heapq.heappop(minHeap)    
        res = []
        # remove the elements from Heap and append to the result and return it
        while minHeap:
            res.insert(0,heapq.heappop(minHeap).string)
        
        return res


# Your AutocompleteSystem object will be instantiated and called as such:
# obj = AutocompleteSystem(sentences, times)
# param_1 = obj.input(c)

File: AutocompleteBrute.java
class AutocompleteSystem {
    //tc-o(n) + map(o(n))
    HashMap<String, Integer> map;
    StringBuilder inputs;
    public AutocompleteSystem(String[] sentences, int[] times) {
        this.map = new HashMap<>();
        this.inputs = new StringBuilder();

        for(int i=0;i<sentences.length;i++)
        {
            String sentence = sentences[i];
            map.put(sentence, map.getOrDefault(sentence,0)+times[i]);
        }

    }
    
    public List<String> input(char c) {
        if(c == '#')
        {
            String search = inputs.toString();
            map.put(search, map.getOrDefault(search,0)+1);
            inputs = new StringBuilder();
            return new ArrayList<>();

        }
        inputs.append(c);
        PriorityQueue<String> pq = new PriorityQueue<>((a,b)->{
            if(map.get(a) == map.get(b))//when freq are equal smaller letter will have precendence
            {
                return b.compareTo(a);// 
                /*String myStr1 = "ir";
String myStr2 = "is";
System.out.println(myStr2.compareTo(myStr1)); s2.comapres1 retruns 1 , s1comapres2 retruns -1
  String myStr1 = "ii";//1st
String myStr2 = "im";//4nd
System.out.println(myStr1.compareTo(myStr2));//i-m = -4*/
            }
            else
            {
                return map.get(a) - map.get(b);//string with lesser freq should be on the top
            }
         });
         String prefix = inputs.toString();
         for(String key : map.keySet())
         {
             if(key.startsWith(prefix))
             {
                 pq.add(key);
             }
             if(pq.size()>3)
             {
               pq.poll();
             }
         }
        List<String> result = new ArrayList<>();
        while(!pq.isEmpty())
        {
            result.add(0,pq.poll());
        }
        return result;
    }
}

/**
 * Your AutocompleteSystem object will be instantiated and called as such:
 * AutocompleteSystem obj = new AutocompleteSystem(sentences, times);
 * List<String> param_1 = obj.input(c);
 */